### PR TITLE
Improve a flow of loading devices and xcode detection

### PR DESCRIPTION
**Pull request by filip131311**
this PR addresses a few issues with the loading devices  flow: 
- `checkIfCLIInstalled` no longer logs errors when returning false instead just logs with debug level, this change makes the logs more alined with expected results for setups without xcode
- initial device load is performed with force, as xcode installation could change between session of the ide
- initial loading screen no longer "jumps" between states 
- removed `DYLD_FRAMEWORK_PATH` as it is no longer needed
- `trySelectingInitialDevice()` creates "devicesChanged" listeners before selectingInitial device is complete, so it catches cases in which `lastDeviceId` points to the device that no longer exists (e.g. because of xcode update)

### CODE REVIEW

File: packages/vscode-extension/src/dependency/DependencyManager.ts [hunk 1]
```diff
@@ -428,9 +425,12 @@ export class DependencyManager implements Disposable {
 export async function checkIfCLIInstalled(cmd: string, options: Record<string, unknown> = {}) {
   try {
     // We are not checking the stderr here, because some of the CLIs put the warnings there.
-    const { stdout } = await command(cmd, { encoding: "utf8", ...options });
-    return !!stdout.length;
+    const { stdout } = await command(cmd, { encoding: "utf8", quiet: true, ...options });
```
Comment (jakub-gonet) [inline packages/vscode-extension/src/dependency/DependencyManager.ts:428]: `quiet` name needs to change eventually to something like `printStdErr`, I'd assume if quiet is true we only get status code on exit.

File: packages/vscode-extension/src/dependency/DependencyManager.ts [hunk 2]
```diff
@@ -428,9 +425,12 @@ export class DependencyManager implements Disposable {
...
+    const result = !!stdout.length;
```
Comment (jakub-gonet) [inline packages/vscode-extension/src/dependency/DependencyManager.ts:429]: Please be more explicit:
```
const hasOutput = stdout.lenght > 0;
```

File: packages/vscode-extension/src/dependency/DependencyManager.ts [hunk 3]
```diff
@@ -428,9 +425,12 @@ export class DependencyManager implements Disposable {
...
+    Logger.debug(`CLI: ${cmd} ${result ? "" : "not"} installed `);
```
Comment (jakub-gonet) [inline packages/vscode-extension/src/dependency/DependencyManager.ts:430]: Nit: "CLI: " prefix is unnecessary.

File: packages/vscode-extension/src/devices/DeviceManager.ts [hunk 4]
```diff
@@ -125,11 +128,11 @@ export class DeviceManager implements DeviceManagerInterface {
     return getAvailableIosRuntimes();
   }
 
-  public async listAllDevices() {
+  public async listAllDevices(force?: boolean) {
```
Comment (jakub-gonet) [inline packages/vscode-extension/src/devices/DeviceManager.ts:131]: Maybe `forceReload`?

File: packages/vscode-extension/src/webview/views/PreviewView.tsx [hunk 5]
```diff
@@ -154,9 +167,10 @@ function PreviewView() {
           onZoomChanged={onZoomChanged}
         />
       ) : (
-        <div className="missing-device-filler">
-          {devicesNotFound ? <DevicesNotFoundView /> : <VSCodeProgressRing />}
-        </div>
+        <LoadingComponent
+          finishedInitialLoad={finishedInitialLoad}
```
Comment (jakub-gonet) [inline packages/vscode-extension/src/webview/views/PreviewView.tsx:171]: Why we're checking for finishedInitialLoad twice in the ternary and in the component?

File: packages/vscode-extension/src/dependency/DependencyManager.ts (hunk 3 above)
Comment (filip131311) [reply packages/vscode-extension/src/dependency/DependencyManager.ts:430]: I think it is useful to have prefixes that allow to serach what part of our code generated the log. We have. a growing code base and most of that log is generated dynamically so searching for it would be hard without it.

File: packages/vscode-extension/src/webview/views/PreviewView.tsx (hunk 5 above)
Comment (filip131311) [reply packages/vscode-extension/src/webview/views/PreviewView.tsx:171]: there is more then one reason for rendering the loading component, but not all of them behave the same.

File: packages/vscode-extension/src/dependency/DependencyManager.ts (hunk 2 above)
Comment (filip131311) [inline packages/vscode-extension/src/dependency/DependencyManager.ts:429]:
```suggestion
    const result = stdout.length > 0;
```

File: packages/vscode-extension/src/devices/DeviceManager.ts (hunk 4 above)
Comment (filip131311) [inline packages/vscode-extension/src/devices/DeviceManager.ts:131]:
```suggestion
  public async listAllDevices(forceReload?: boolean) {
   const devices = extensionContext.globalState.get(DEVICE_LIST_CACHE_KEY) as
      | DeviceInfo[]
      | undefined;
    if (devices && !forceReload) {
```

File: packages/vscode-extension/src/devices/DeviceManager.ts [hunk 6]
```diff
@@ -104,13 +105,15 @@ export class DeviceManager implements DeviceManagerInterface {
...
     });
-    const simulators =
-      Platform.OS === "macos"
-        ? listSimulators().catch((e) => {
-            Logger.error("Error fetching simulators", e);
-            return [];
-          })
-        : Promise.resolve([]);
+
+    const shouldLoadSimulators = Platform.OS === "macos" && (await checkXcodeExists());
```
Comment (kmagiera) [inline packages/vscode-extension/src/devices/DeviceManager.ts:109]: Maybe we can add additional log here when checkXcodeExists fails. For example "Couldn't list iOS simulators as XCode installation wasn't found" – this will allow us to track issues in case those checks aren't accurate

File: packages/vscode-extension/src/webview/providers/ProjectProvider.tsx [hunk 7]
```diff
@@ -26,6 +26,7 @@ const ProjectContext = createContext<ProjectContextProps>({
       longitude: 19.965474,
       isDisabled: false,
     },
+    locale: "en_US",
```
Comment (kmagiera) [inline packages/vscode-extension/src/webview/providers/ProjectProvider.tsx:29]: why is this change here?

File: packages/vscode-extension/src/devices/DeviceManager.ts [hunk 8]
```diff
@@ -125,11 +128,11 @@ export class DeviceManager implements DeviceManagerInterface {
     return getAvailableIosRuntimes();
   }
 
-  public async listAllDevices() {
+  public async listAllDevices(forceReload?: boolean) {
     const devices = extensionContext.globalState.get(DEVICE_LIST_CACHE_KEY) as
       | DeviceInfo[]
       | undefined;
-    if (devices) {
+    if (devices && !forceReload) {
```
Comment (kmagiera) [inline packages/vscode-extension/src/devices/DeviceManager.ts:135]: this isn't accurate as we *always* load the devices. The difference is that we return the updated list or let it update asynchronously. I'm a bit hesitant about this change as from what I remember, we added device caching as it was slowing down the initial startup significantly. As it isn't documented in PR description I wonder why you're changing this behavior?

File: packages/vscode-extension/src/webview/providers/ProjectProvider.tsx (hunk 7 above)
Comment (filip131311) [reply packages/vscode-extension/src/webview/providers/ProjectProvider.tsx:29]: It was missed before and ts was not happy, it does not effect the webview much at all as it never uses default context.

File: packages/vscode-extension/src/devices/DeviceManager.ts (hunk 8 above)
Comment (filip131311) [reply packages/vscode-extension/src/devices/DeviceManager.ts:140]: When user had an iPhone selected when they last used IDE, but something happend to xcode (e.g. user started using beta but not updated the `xcode-select`)  application would select an iPhone anyway as it is the most recent device and would get stuck on loading screen. 

What dou you think about `forceLoad` instead of `forceReload`

Comment (kmagiera) [reply packages/vscode-extension/src/devices/DeviceManager.ts:140]: I think we should optimize for the most common case in which people don't change their xcode location. If there are cases when it happens, I understand they will get an error when trying to load the simulator, and by that time we will have the list refetched. If that's not the case we should adjust the logic such that the worst thing that can happen if you mess up the simjulator cache or xcode is that you get an error loading the simulator instead of slowing down all the time to interact for everyone else.

File: packages/vscode-extension/src/devices/DeviceManager.ts [hunk 9]
```diff
@@ -104,13 +105,20 @@ export class DeviceManager implements DeviceManagerInterface {
...
-      Platform.OS === "macos"
-        ? listSimulators().catch((e) => {
-            Logger.error("Error fetching simulators", e);
-            return [];
-          })
-        : Promise.resolve([]);
+
+    let shouldLoadSimulators = Platform.OS === "macos";
+
+    if (!(await checkXcodeExists())) {
```
Comment (kmagiera) [inline packages/vscode-extension/src/devices/DeviceManager.ts:111]: do we even want to run those checks if we're not on a Mac? 

Perhaps doing it this way would be cleaner:
```ts
const shouldLoadSimulators = Platform.OS === "macos" && (await checkXcodeExists());
```

Then you can move the log to be printed in `checkXcodeExists`

File: packages/vscode-extension/src/project/project.ts [hunk 10]
```diff
@@ -201,13 +200,21 @@ export class Project
...
+        return;
+      }
+      if (await selectInitialDevice(newDevices)) {
+        this.deviceManager.removeListener("devicesChanged", listener);
+      }
+    };
+    this.deviceManager.addListener("devicesChanged", listener);
+
+    if (await selectInitialDevicePromise) {
+      this.deviceManager.removeListener("devicesChanged", listener);
```
Comment (kmagiera) [inline packages/vscode-extension/src/project/project.ts:217]: This code doesn't do the same thing as the one on the left. In general I'm most afraid of the changes in awaiting that you're adding here. Not saying we shouldn't have them, but they require some explanation in the PR description, such that we can understand why we're changing the await flow.

This part is the most complex change and it introduces some unusual pattern (that we, albeit, use in other places but typically with some explanation) where we create promise and then await inside of an arrow function.

I think that given this method already awaits `selectDevice` (which wasn't the case before this change) , we can just add the listener before calling `selectInitialDevice` ?

**Review by kmagiera — COMMENTED** [review]
Looks mostly good, left a few comments inline

File: packages/vscode-extension/src/devices/DeviceManager.ts (hunk 9 above)
Comment (filip131311) [reply packages/vscode-extension/src/devices/DeviceManager.ts:111]: I did this in this way to address your previous comment about adding logging in !checkXcodeExist() scennario

**Review by kmagiera — APPROVED** [review]
lgtm 👍
